test/graphql_test.rb

```diff
@@ -374,4 +374,35 @@ def test_mutation_batch_subselection_execution
     assert_equal expected, result
     assert_equal ["Product/1,2", "Product/2,3"], queries
   end
+
+  class MyCustomExecutor < GraphQL::Batch::Executor
+    attr_reader :loaders_executed
+
+    def initialize
+      @loaders_executed = []
+      super
+    end
+
+    def around_promise_callbacks(loader)
+      @loaders_executed << loader
```

Comment:
This is called each time a promise is fulfilled, not for each loader being executed.  E.g. if 4 keys are being loaded by a loader, then this will be called for all 4 keys rather than the one load.

Reply:
> This is called each time a promise is fulfilled, not for each loader being executed.

Thanks for the clarification!

It looks like `around_promise_callbacks` is also invoked via `Loader#reject`. So that's twice per field, even if the field is fulfilled. `reject` is called even when the loader finds a result, via `check_for_broken_promises`. This explains the double logging in my current implementation.

```
lib/graphql/batch/loader.rb:78:in `reject'
lib/graphql/batch/loader.rb:131:in `block in check_for_broken_promises'
lib/graphql/batch/loader.rb:130:in `each'
lib/graphql/batch/loader.rb:130:in `check_for_broken_promises'
lib/graphql/batch/loader.rb:48:in `resolve'
lib/graphql/batch/executor.rb:34:in `resolve'
lib/graphql/batch/loader.rb:56:in `wait'
lib/promise.rb:129:in `wait'
lib/promise.rb:129:in `wait'
lib/promise.rb:129:in `wait'
lib/promise.rb:80:in `sync'
lib/graphql/field.rb:309:in `public_send'
```

Reply:
I was able to  around the double-logging by resetting `@start_time` to nil in `around_promise_callbacks`, like so:

```ruby
module GraphModel
  class TimedBatchExecutor < GraphQL::Batch::Executor
 
    # etc.
    
    def resolve(loader)
      @start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      super
    end

    def around_promise_callbacks(loader)
      return super unless @start_time

      end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      TIMINGS[loader.loader_key] ||= []
      TIMINGS[loader.loader_key] << end_time - @start_time

      @start_time = nil

      super
    end
  end
end
```

This prevents the second call to `around_promise_callbacks` (via Executor#resolve -> Loader#resolve, #check_for_broken_promises, #reject, #finish_resolve) from logging the time between the original `@start_time` and the time it runs again.